src/openforms/forms/validators.py

```diff
@@ -92,3 +99,77 @@ def validate_template_expressions(configuration: JSONObject) -> None:
         all_errors.append(err)
 
     raise ValidationError(all_errors)
+
+
+@dataclass
+class FakeFormDefinition:
+    configuration: JSONObject
+    name: str = ""
+
+
+def validate_no_duplicate_keys(
+    configuration: JSONObject,
+) -> None:
+    form_definition = FakeFormDefinition(configuration=configuration)
+    validate_no_duplicate_keys_across_steps(form_definition, other_form_definitions=[])
+
+
+def validate_no_duplicate_keys_across_steps(
+    current_form_definition, other_form_definitions
+):
+    """
+    Validate that there are no duplicate keys in a configuration.
+
+    This is important in the case of repeating groups, because no variables are created for the components inside the
+    repeating group, so there is no database constraint for the keys to be unique. However, non-unique keys cause
+    problems in the case of file components.
+
+    #TODO This is a bandaid, this problem should be fixed properly in #2728
+    """
+    component_path_map = defaultdict(list)
+    duplicate_keys = []
+    for form_definition in [current_form_definition] + other_form_definitions:
+        for configuration_path, component in flatten_by_path(
+            form_definition.configuration
+        ).items():
+            key = component["key"]
+            component_path_map[key].append((configuration_path, form_definition))
+
+            if key not in duplicate_keys and len(component_path_map[key]) > 1:
+                duplicate_keys.append(key)
+
+    if not duplicate_keys:
+        return
+
+    # Get readable path for duplicate components
+    errors = []
+    for duplicate_key in duplicate_keys:
+        # If the duplicates are all in other steps, don't raise it here from this step
+        if all(
+            [
+                form_definition != current_form_definition
+                for path, form_definition in component_path_map[duplicate_key]
+            ]
+        ):
+            continue
+
+        # The readable path is in format "prefix > parent component > child component"
+        readable_paths = [
+            get_readable_path_from_configuration_path(
+                form_definition.configuration, configuration_path, form_definition.name
+            )
+            for configuration_path, form_definition in component_path_map[duplicate_key]
+        ]
+        errors.append(
+            _('"{duplicate_key}" (in {paths})').format(
+                duplicate_key=duplicate_key, paths=", ".join(readable_paths)
+            )
+        )
+
+    if errors:
+        raise ValidationError(
+            format_lazy(
+                "Detected duplicate keys in configuration: {errors}",
+                errors=get_text_list(errors, ", "),
+            )
```

Comment:
I'm not sure if the `format_lazy` call is appropriate here? And shouldn't the "Detected duplicate keys..." be marked translatable too?